avm/src/vm/opcode.rs

```diff
@@ -3076,18 +3117,86 @@ pub static OPCODES: Lazy<HashMap<i64, ByteOpcode>> = Lazy::new(|| {
   }
   io!(httplsn => fn(_args, hand_mem) {
     Box::pin(async move {
-      let port_num = Program::global().http_port;
-      let addr = SocketAddr::from(([0, 0, 0, 0], port_num));
-      let make_svc = make_service_fn(|_conn| async { Ok::<_, Infallible>(service_fn(http_listener)) });
-
-      let bind = Server::try_bind(&addr);
-      match bind {
-        Ok(server) => {
-          let server = server.serve(make_svc);
+      match &Program::global().http_config {
+        HttpType::HTTP(http) => {
+          let port_num = http.port;
+          let addr = SocketAddr::from(([0, 0, 0, 0], port_num));
+          let make_svc = make_service_fn(|_conn| async { Ok::<_, Infallible>(service_fn(http_listener)) });
+
+          let bind = Server::try_bind(&addr);
+          match bind {
+            Ok(server) => {
+              let server = server.serve(make_svc);
+              tokio::spawn(async move { server.await });
+              println!("HTTP server listening on port {}", port_num);
+            },
+            Err(ee) => eprintln!("HTTP server failed to listen on port {}: {}", port_num, ee),
+          }
+        },
+        HttpType::HTTPS(https) => {
+          let port_num = https.port;
+          let addr = SocketAddr::from(([0, 0, 0, 0], port_num));
+          let tls_cfg = {
+            let certs = pemfile::certs(
+              &mut BufReader::new(base64::decode(https.cert_b64.as_str()).unwrap().as_slice())
+            );
+            let certs = match certs {
+              Ok(certs) => certs,
+              Err(_) => panic!("Failed to load certificate"),
+            };
+            eprintln!("{:?}", certs);
+            let key = {
+              let keys = pemfile::pkcs8_private_keys(
+                &mut BufReader::new(
+                  base64::decode(https.priv_key_b64.as_str()).unwrap().as_slice()
+                )
+              );
+              let keys = match keys {
+                Ok(keys) => keys,
+                Err(_) => panic!("Failed to load private key"),
+              };
+              if keys.len() != 1 {
+                eprintln!("{:?}", keys);
+                panic!("Expected a single private key");
+              }
+              keys[0].clone()
+            };
+            let mut cfg = rustls::ServerConfig::new(rustls::NoClientAuth::new());
+            cfg.set_single_cert(certs, key)
+              .map_err(|e| panic!(e)).unwrap();
```

Comment:
this can just be `unwrap`

Reply:
Yeah, that was my refactoring of the example and I forgot to clean this up. It used to be filled with `?` and so automatically returned an error Result here.